Auth cloud mail UI

kalash9747 opened the pull request:
Сделал Ui тесты облака с использованием БД 
![image](https://user-images.githubusercontent.com/82981820/170108278-04b76b69-b71c-4236-a742-521bf7617fbf.png)
![image](https://user-images.githubusercontent.com/82981820/170108364-771df767-9d42-4db9-a550-2debae3e16ce.png)
![image](https://user-images.githubusercontent.com/82981820/170108443-9ab85f07-5ec9-4b6c-b44d-d46ca7beea33.png)

kalash9747 committed:
add check ui authorization cloud mail

kalash9747 committed:
not see this commit

kalash9747 committed:
file comparison is working

kalash9747 committed:
Merge branch 'authorizationCloudMail' into authCloudMailUi

# Conflicts:
#	cloudmail/build.gradle
#	util/build.gradle

kalash9747 committed:
add HomePageTest
add TestRunner
add selenide dependencies

kalash9747 committed:
renamed shouldContainedText -> shouldContainText
corrected spelling in comments

kalash9747 committed:
add removeFile
add ByAttribute

dKusakin commented on cloudmail/build.gradle:
```diff
@@ -14,15 +14,20 @@ allure {
     testImplementation jacksonCore
     testImplementation jacksonDataBind
 
+    testRuntimeOnly browserUpProxyCore
```
`selenide-proxy` не устроил? Да и нужен ли он, если ты `FOLDER` заюзал?

kalash9747 replied:
убрал

dKusakin commented on cloudmail/src/main/java/models/dbModels/CloudFileInfo.java:
```diff
@@ -0,0 +1,57 @@
+import static java.lang.System.getProperty;
+
+/**
+ * Представление фойла в базе данных
```
ф**а**йла

kalash9747 replied:
поправил

dKusakin commented on cloudmail/src/main/java/models/dbModels/CloudFileInfo.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public String getNameWithExt() {
+        return name + contentextension;
```
Ты расширения в БД сразу с точкой хранишь?

kalash9747 replied:
убрал точку  и из БД и в проекте

dKusakin commented on cloudmail/src/main/java/models/dbModels/CloudFileInfo.java:
```diff
@@ -0,0 +1,57 @@
+    public File toTempFile() {
+        File file = new File(getProperty("java.io.tmpdir"), this.getNameWithExt());
+        file.deleteOnExit();
+        try (FileOutputStream fos = new FileOutputStream(file.getPath())) {
```
Просто `file` в конструктор чем не катит?

kalash9747 replied:
сделал

dKusakin commented on util/src/main/java/annotations/Column.java:
```diff
@@ -0,0 +1,9 @@
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Column {
```
Что-то нигде не нашел использование, нужна?

kalash9747 replied:
затесалась после экспериментов, убрал

dKusakin commented on cloudmail/src/main/java/pages/AuthorizationFrame.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Step("Нажать кнопку 'Ввести пароль'")
+    public AuthorizationFrame nextButtonClick() {
+        nextButton().shouldBe(Condition.visible).click();
```
Ты знаешь, что делать)

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,133 @@
+    public HomePage cellContentVisible(String name, String extension) {
+        SelenideElement cell = cell(name, extension).shouldBe(visible);
+        cell.$(byText(name)).shouldBe(visible);
+        cell.$(byText(extension.replace(".", ""))).shouldBe(visible);
```
Похоже, и правда с точкой. А зачем? Можно ж было б без неё в БД хранить, т.к. смысл в этой точке?

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public SelenideElement cell(String name, String extension) {
+        return cells()
+                .filterBy(matchText(name))
```
Тебе именно регулярка нужна?
Простой `text` будет тоже по частичному соответствию искать (там `contains` под капотом)

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,133 @@
+    public File downloadFile(CloudFileInfo cloudFile) {
+        cell(cloudFile.getName(), cloudFile.getContentextension()).click();
+        File file = null;
+        downloadButton().click();
```
А зачем этот клик? Разве `download` ниже за тебя не кликнет?

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,133 @@
+        try {
+            file = downloadButton().download(10000, withName(cloudFile.getNameWithExt()));
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
А есть смысл продолжать? Кидал бы тут `AssertionFailedError` и всего делов

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,103 @@
+    @BeforeEach
+    void authorization() {
+        homePage = login(user).userWidgetCheck(user);
+        fileFromDB = getAllCloudFilesFromDB().get(0).setName(currentTimeMillis() + "");
```
Ээ, а чего `get(0)` ? Всё время будет первую брать? Не, надо случайно тыкать

kalash9747 replied:
верно, сделал

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,103 @@
+public class HomePageTest extends TestRunner {
+    private static final User user = getUser(admin);
+    private static HomePage homePage;
+    private static CloudFileInfo fileFromDB;
```
Есть тут смысл от статика?

kalash9747 replied:
убрал, мои грабли ходят по пятам

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,103 @@
+    @Test
+    void downloadFileCheck() {
+        File fileFromCloud = homePage
+                .uploadFile(fileFromDB.toTempFile())
```
Я изначально задумывал эти кейсы как один выгружает, а второй загружает. Теперь понимаю, что нужен ещё третий, который удаляет)
Есть мысль, как это реализовать с помощью БД и `assume`. Например, при выгрузке получать доступные файлы из БД, понять, каких нет на UI, и из оставшихся выбрать какой-нибудь для выгрузки. В `assume` поверять, что ещё есть что выгружать. С остальными аналогично. По идее, так это будет более атомарно, но надо поприседать. Да и не уверен, что оно надо

kalash9747 replied:
Вопрос открытый

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    private byte[] fileToByteArray(File file) {
+        try {
+            return readFileToByteArray(file);
```
Это же 11-я джава! Тут у `InputStream` (а, значит, и у `FileInputStream`) есть прекрасный метод `readAllBytes()`

kalash9747 replied:
сделал

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,103 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static util.Authorization.login;
+
+public class HomePageTest extends TestRunner {
```
Лан, в QE пропущу тут наследование, в SQE, скорее всего, уже нет)

kalash9747 replied:
в SQE так и не буду делать)

dKusakin reviewed (changes requested):
Комменты написал + бд мне никто не дал, чтобы потыкать :)

kalash9747 committed:
deleted unused dependencies
deleted Column
added throw new AssertionFailedError
fixed comment
changed getNameWithExt() in CloudFileInfo

kalash9747 commented:
все поправил, кроме тестов для remove, пока вопрос на обсуждении

kalash9747 committed:
add removeAllFilesNotContainedInDBFromCloud
removed the dependency of downloading from uploading

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,158 @@
+            return downloadButton().download(10000, withName(getNameWithExtFromCell(cell)));
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+            throw new AssertionFailedError("Не удалось скачать файл");
```
Всё же вторым аргументом `e` неплохо бы подвязать. Так `StackTrace` туда передастся оригинальный тоже. Ну и вывод выше тебе будет ни к чему

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,131 @@
+     */
+    private byte[] fileToByteArray(File file) {
+        try {
+            return new FileInputStream(file).readAllBytes();
```
А закрывать кто будет? Думаю, ты знаешь, какой конструкцией

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,131 @@
+    @Owner("Калашников Владислав Александрович")
+    @Test
+    void downloadFileCheck() {
+        File fileFromCloud = homePage.downloadRandomFile();
```
Мы же с тобой говорили, что качаем не случайный файл, а какой-нибудь, соответствующий какому-нибудь файлу из БД по имени. Ты что-то другое тут решил?

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,131 @@
+                .collect(toSet());
+
+        ElementsCollection cells = homePage.cells();
+        for (int i = 0; i < cells.size() - 1; i++) {
```
А если цикл с конца сделать, то мб попроще выйдет?

dKusakin replied:
Вообще не рекомендуют итерировать `ElementsCollection`, так что надо это делать осторожно
Я б, наверное, вытащил эти имена фильтром и потом уже по именам бы удалял. Это так, мысли вслух)

dKusakin reviewed (changes requested):
Закрытие файла и обратная итерация по коллекции

kalash9747 committed:
changed uploadInput
changed downloadFileCheck

dKusakin commented on cloudmail/src/main/java/pages/HomePage.java:
```diff
@@ -0,0 +1,167 @@
+     */
+    public List<String> getFullNamesFromAllCells() {
+        List<String> names = new ArrayList<>();
+        cells().asFixedIterable().forEach(cell -> names.add(getNameWithExtFromCell(cell)));
```
В новой версии селенида этим `Iterable` добавили `stream`)
Можно поглядеть на досуге

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,143 @@
+        for (int i = cells.size() - 1; i >= 0; i--) {
+            if (!filesFromDBNames.contains(homePage.getNameWithExtFromCell(cells.get(i)))) {
+                homePage.removeFile(cells.get(i));
+                cells = homePage.cells();
```
А это зачем? У тебя же коллекция самообновляемая

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,143 @@
+public class HomePageTest extends TestRunner {
+    private final User user = getUser(admin);
+    private List<CloudFileInfo> allFilesFromDB;
+    private CloudFileInfo fileFromDB;
```
Этот парень очень тут нужен?

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,143 @@
+        fileFromDB = allFilesFromDB.stream()
+                .filter(fileInfo -> fileInfo.getNameWithExt().equals(getNameWithoutTimestamp(fileFromCloud.getName())))
+                .findFirst()
+                .orElse(new CloudFileInfo());
```
По-моему, тут логичнее не пустой объект давать, а кидать через `orElseThrow` какой-нибудь `IllegalStateException`, ведь такого, по идее, в штатной ситуации не может быть (имя же изначально сверяли)

dKusakin commented on cloudmail/src/test/java/ui/HomePageTest.java:
```diff
@@ -0,0 +1,143 @@
+                .findFirst()
+                .orElse(new CloudFileInfo());
+        step("Проверить содержимое загруженного файла", () ->
+                assertArrayEquals(fileFromDB.getContentbytes(), fileToByteArray(fileFromCloud),
```
Вот, кстати, из-за предыдущего пустого объекта эта проверка при скачанном пустом файле может быть ложноположительной

kalash9747 committed:
orElse->orElseThrow
removed re-assignment of cells
